src/java.base/share/classes/java/lang/String.java

```diff
@@ -3454,7 +3454,14 @@ public static String join(CharSequence delimiter,
         Objects.requireNonNull(delimiter);
         Objects.requireNonNull(elements);
         var delim = delimiter.toString();
-        var elems = new String[8];
+        String[] elems;
+        long initLen = elements.spliterator().estimateSize();
+        if (initLen <= 8L || initLen >= ArraysSupport.SOFT_MAX_ARRAY_LENGTH) {
+            // This usually means the size of the Iterable cannot be computed easily.
+            elems = new String[8];
+        } else {
+            elems = new String[(int)initLen];
+        }
         int size = 0;
         for (CharSequence cs: elements) {
```

Comment:
I would think you have to locally store the result of `elements.spliterator()` and then use Spliterators::iterator to adapt it back to an iterator.  This should correctly handle [early-binding](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/Spliterator.html#binding) spliterators.

I think in the loop the code should use ArraysSupport.newLength.

Reply:
Note that this also has to handle the case where `si.estimateSize()` returns values close to `Long.MAX_VALUE`:
```java
final var si = elements.spliterator();
final long est = si.estimateSize();
final int length;

final long MIN_LENGTH = 8;
if ((si.hasCharacteristics() & Spliterator.SIZED) != 0) {
	length = (int) Math.clamp(est, MIN_LENGTH, ArraysSupport.SOFT_MAX_ARRAY_LENGTH);
} else if (est >= ArraysSupport.SOFT_MAX_ARRAY_LENGTH - MIN_LENGTH) {
	length = ArraysSupport.SOFT_MAX_ARRAY_LENGTH;
} else {
	length = (int) Math.clamp(est + MIN_LENGTH, MIN_LENGTH, ArraysSupport.SOFT_MAX_ARRAY_LENGTH);
}

var elems = new String[length];
```